rfcs/0000-rda-claims-jsonb.md

```diff
@@ -0,0 +1,341 @@
+- `lastUpdated`: to track last time the record was updated
+- `claim`: entire claim object as JSON
+
+Each of the queryable non-JSON columns has an associated index.
+The `claim` column is not indexed.
+
+When the API receives a request for a claim with a particular MBI a simple query is performed on that column.
+The `claim` value of the matching record is returned as part of the query.
+In memory this JSON is converted into an object graph using an open source library.
+This conversion happens seamlessly using a JPA feature, the `Convert` field annotation.
+
+Code changes are minimal across the BFD code base since the claim data objects are simply converted from JPA entities to Plain Old Java Objects by removing the JPA annotations.
+Two new entity classes are added for the two remaining database tables.
+These have a field for each of the queryable columns plus a field to hold the claim POJO.
+Clients perform queries using the new entity class but then use the returned POJO exactly as they previously used the old JPA entities.
+
+
+### Proposed Solution: Detailed Design
+[Proposed Solution: Detailed Design]: #proposed-solution-detailed-design
+**This section discusses changes specifically for FISS claims.  The changes for MCS claims are directly analagous to the ones for FISS.**
+
+
+The `bfd-model-rda` objects are modified as follows:
+
+- All of the JPA annotations are removed from the existing JPA entity classes (PreAdjFissClaim, PreAdjFissPayer, etc).
+- All of the static inner classes for composite keys are removed from the existing entity classes as well.
+- Redundant fields are removed from the detail objects (e.g. `dcn` is no longer needed in `PreAdjFissPayer`).
+- One new JPA entity class is added for each claim type (`PreAdjFissClaimJson` and `PreAdjMcsClaimJson`).  Details below.
+- An abstract base class implementing the JPA `AttributeConverter` interface is added.  Details below.
+- A concrete implementation of this base class is added for each claim type (`PreAdjFissClaimConverter` and `PreAdjMcsClaimConverter`).
+
+The database schema is simplified to a single table for each claim type:
+
+````sql
+create table "pre_adj"."FissClaimsJson" (
+    "dcn"            varchar(23)   not null,
+    "mbi"            varchar(13),
+    "mbiHash"        varchar(64),
+    "stmtCovToDate"  date,
+    "sequenceNumber" bigint        not null,
+    "lastUpdated"    timestamp with time zone,
+    "claim"          ${type.jsonb} not null,
+    constraint "FissClaimsJson_pkey" primary key ("dcn")
+);
+````
+The schema also contains one index for each of the queryable columns.
+
+Note the template macro for the `claim` column type.
+This macro is used since the type of the column will be different in postgresql vs HSQLDB:
+- In postgresql the `JSONB` type is used to allow use of JSON path querying for ad-hoc queries.
+- In HSQLDB `varchar(max)` is used since `JSONB` is not supported for that database.
```

Comment:
Will we ever exceed varchar(max) size in the jsonb payload...I think hsql maxes at 8k.

Reply:
Good catch @mackec1 .  I actually used `LONGVARCHAR` in the prototype, not `varchar(max)`.  My mistake.  `LONGVARCHAR` defaults to 16mb but if could be be configured to use a clob in which case it could be gigabytes.

Reply:
I have fixed the RFC.